### ISSUE TITLE
TestCafe Issues when using IE11

### ISSUE THREAD

**Issue by nabond240** [DevExpress/testcafe#2750]
### Are you requesting a feature or reporting a bug?
Reporting a bug

### What is the current behavior?
An "Invalid character" error is displayed when entering text into a field on an IE11 browser window using .typeText().

### What is the expected behavior?
No error should be displayed (i.e. when using Chrome or Firefox)

### How would you reproduce the current behavior (if this is a bug)?
Created a simple test.js file to automate the following:
1. Open https://login.yahoo.com/
2. Enter string "temporary" into the Enter your email field.

#### Provide the test code and the tested page URL (if applicable)

Tested page URL:

Test code

```js

```

### Specify your

* operating system:  Windows 10
* testcafe version:  0.21.1
* node.js version:  8.11.3

**Comment by AndreyBelym** [comment]
Hi @nabond240, I can't reproduce it, could you please try to run my version and post your version here?
```js
fixture `Yahoo`.page `login.yahoo.com`;

test(`Login`, t => t.typeText('#login-username', 'temporary').wait(4000))
```

**Comment by nabond240** [comment]
I've copied your version and tried it out, but I'm still getting the same result.  Please see "yahoo error.txt" below.  Are there any settings that need to be changed in the IE browser that may prevent this from occurring?

I added a "t.maximizeWindow();", and at times it seems to work the 1st time, but executing the same exact code again will fail every time after.  Please see "double execution.txt" below.  Let me know if you need anything else.

[yahoo error.txt](https://github.com/DevExpress/testcafe/files/2306253/yahoo.error.txt)

[double execution.txt](https://github.com/DevExpress/testcafe/files/2306297/double.execution.txt)

**Comment by nabond240** [comment]
It may be possible that it has to do with the connection to the internet that is established.  I noticed the issue only when connecting to a private network.  This issue does not occur when connected to a public network.

I am still not sure why it works perfectly for Chrome and Firefox, and I am curious why it's reporting an "Invalid character" error message if it is related to my connection.

**Comment by link89** [comment]
[testcafe-ie11-error.txt](https://github.com/DevExpress/testcafe/files/2369046/testcafe-ie11-error.txt)
@AndreyBelym Here is my result run the fixture you provide.
testcafe version: 0.22.0
ie version: 11.0.22

**Comment by link89** [comment]
I think it has something to do with web site under test. Our web site suffers same problem, but following code for todomvc works well.
```javascript
fixture `TodoMVC`.page `http://todomvc.com/examples/react/#/`;
test(`Login`, t => t.typeText('.new-todo', 'temporary').wait(4000))
```

**Comment by link89** [comment]
@AndreyBelym Another critical problem on IE11 is, the url generate by testcafe is incorrect! For example, the heartbeat url on IE11 is:
`http://10.32.60.135:54899/Oi7Th2mYy/http://todomvc.com/examples/react/undefined/browser/heartbeat/j2d8olI`
The expected one should be
`http://10.32.60.135:54899/Oi7Th2mYy/browser/heartbeat/j2d8olI`
It seems to me like the problem of location.origin is undefined on IE11, and testcafe or testcafe-hammerhead didn't handle it correctly.

**Comment by AndreyBelym** [comment]
@miherlosev, @LavrovArtem, looks like the hash symbol breaks URL parsing when used as a path element.

**Comment by AndreyBelym** [comment]
@link89, @nabond240, could you please install the `testcafe@0.22.1-alpha.1` build and try it in your environment? We've implemented printing stack traces of JS errors in browsers, so it can give precise location where the error is created.

**Comment by AndreyBelym** [comment]
Could you please check that your version of IE11 doesn't have the `origin` property in the `location` object?
![image](https://user-images.githubusercontent.com/4479386/45409726-a972bd80-b678-11e8-9fe4-b9b2ab01092c.png)

**Comment by link89** [comment]
Here it is.
<img width="1050" alt="10_32_35_76_25001" src="https://user-images.githubusercontent.com/3314130/45410740-0c7a4980-b6a5-11e8-9c1f-3ce7d557292f.png">

**Comment by nabond240** [comment]
> I think it has something to do with web site under test. Our web site suffers same problem, but following code for todomvc works well.
> 
> ```js
> fixture `TodoMVC`.page `http://todomvc.com/examples/react/#/`;
> test(`Login`, t => t.typeText('.new-todo', 'temporary').wait(4000))
> ```

That is correct.  This issue doesn't happen for all websites.  The TodoMVC page works correctly for me as well, and Google works. So far I've seen the issue on our web site and the login.yahoo.com page.

**Comment by nabond240** [comment]
> @link89, @nabond240, could you please install the `testcafe@0.22.1-alpha.1` build and try it in your environment? We've implemented printing stack traces of JS errors in browsers, so it can give precise location where the error is created.

@AndreyBelym I will give this a shot soon and let you know the results.

**Comment by nabond240** [comment]
@AndreyBelym I tried the npm install testcafe@0.22.1-alpha.1 on my local machine and I received an error message.  Please see attached debug log.

[2018-09-14T00_47_24_794Z-debug.log](https://github.com/DevExpress/testcafe/files/2381466/2018-09-14T00_47_24_794Z-debug.log)

**Comment by nabond240** [comment]
@link89 I'm just curious, are you going through a proxy as well when reproducing this issue?  Like I mentioned above, it seems to work fine at home, but when I'm using my job's connection, that's when I'm seeing this issue.

**Comment by link89** [comment]
@nabond240 For me it's the problem of IE11 version. Old version of IE11's location.origin is undefined. Do you use same version of IE11 when you try it at home?

**Comment by AndreyBelym** [comment]
@nabond240, I don't have a proxy in my network. From what I've got, it's a bug of older IE11 versions as @link89 said. Could you please try to update your Internet Explorer installation?

**Comment by nabond240** [comment]
@link89 I'm using the same PC both at work and at home.  What version of IE did you update to for it to work?

@link89, @AndreyBelym I'm using Windows10, so it should update IE automatically.  I double checked and downloaded the available windows updates, restarted my PC, and I'm still noticing the issue.  I did check the location.origin mentioned above and it's undefined for me.  I'm not sure if it matters, but currently I'm using the LTS version of Windows

![test](https://user-images.githubusercontent.com/42553847/45628377-ef0cfd00-ba61-11e8-9352-663dd3e4391f.jpg)

**Comment by link89** [comment]
@nabond240 You should update your ie11

@AndreyBelym By the way, here is another IE11/Edge problem I found when I tried testcafe@0.22.1-alpha.1: https://github.com/DevExpress/testcafe-react-selectors/issues/119#issuecomment-421897436

**Comment by AndreyBelym** [comment]
@nabond240, I've reproduced the issue with IE 11.0.27 on Windows 10.10240. Looks like IE11 on Windows 10.10240 misses the `location.origin` fix. I will try to update IE to the latest version available, and then merge you PR if it is still reproducible.

@link89, DevExpress/testcafe-react-selectors#119 is a different issue, introduced by [this commit](https://github.com/DevExpress/testcafe-hammerhead/commit/e6e5877ee221780d5717976e37cfa53217a1547f) in `testcafe-hammerhead`. I will investigate your original issue from `testcafe-react-selectors` and open a new issue in the `testcafe-hammerhead` repository.

**Comment by nabond240** [comment]
@AndreyBelym Thanks!  Yea, the Windows version just says "Windows 10 Enterprise 2015 LTSB" for me.  It's a version that prevents you from using MS Edge, and there might be some restrictions on what you can do with IE.

**Comment by miherlosev** [comment]
@nabond240 @link89

The fix was published in `testcafe@0.23.0`.
